File: Advent of Code/2022/day 1/solution.dart
import 'dart:convert';
import 'dart:io';

void part_1() async {
  List<String> file_contents = [];
  int max_elf_calories = 0;

  // Read the input file
  var file_path = "./input.txt";
  final file = File(file_path);

  if (await file.exists()) {
    final linesStream = file.openRead();
    await for (var line
        in linesStream.transform(Utf8Decoder()).transform(LineSplitter())) {
      file_contents.add(line);
    }
  }

  int current_calories_total = 0;

  // Loop through the calories until we reach a blank line
  for (var i = 0; i < file_contents.length; i++) {
    if (file_contents[i].length == 0) {
      if (current_calories_total > max_elf_calories) {
        max_elf_calories = current_calories_total;
      }
      current_calories_total = 0;
    } else {
      current_calories_total += int.parse(file_contents[i]);
    }
  }

  print(max_elf_calories);
}

void part_2() async {
  List<String> file_contents = [];

  // Read the input file
  var file_path = "./input.txt";
  final file = File(file_path);

  if (await file.exists()) {
    final linesStream = file.openRead();
    await for (var line
        in linesStream.transform(Utf8Decoder()).transform(LineSplitter())) {
      file_contents.add(line);
    }
  }

  List<int> elf_calories = [];
  int current_calories_total = 0;

  // Loop through the calories until we reach a blank line
  for (var i = 0; i < file_contents.length; i++) {
    if (file_contents[i].length == 0) {
      elf_calories.add(current_calories_total);
      current_calories_total = 0;
    } else {
      current_calories_total += int.parse(file_contents[i]);
    }
  }

  elf_calories.sort((b, a) => a.compareTo(b));
  print(elf_calories.sublist(0, 3).reduce((a, b) => a + b));
}

void main() {
  part_1();
  part_2();
}

File: Advent of Code/2022/day 1/solution.ts
import * as fs from "fs";

function part_1() {
    const words = fs.readFileSync('./input.txt', 'utf-8');
    const words_array = words.split('\n');

    let max_elf_calories: number = 0;
    let current_calories_total: number = 0;

    for (let index = 0; index < words_array.length; index++) {
        const element = words_array[index];

        /* If the current line is a blank newline then we've reached a
            new elf so we check if `current_calories_total` is greater
            than `max_elf_calories`. If it is then set `max_elf_calories`
            to `current_calories_total`. Then set `current_calories_total` to 0.
        */
        if (element == '') {
            if (current_calories_total > max_elf_calories) {
                max_elf_calories = current_calories_total;
            }
            current_calories_total = 0;
        } else {
            current_calories_total += Number(element);
        }
    }

    console.log(max_elf_calories);
}


function part_2() {
    const words = fs.readFileSync('./input.txt', 'utf-8');
    const words_array = words.split('\n');

    let max_elf_calories: number[] = [];
    let current_calories_total: number = 0;

    for (let index = 0; index < words_array.length; index++) {
        const element = words_array[index];

        /* If the current line is a blank newline then we've reached a
            new elf so we check if `current_calories_total` is greater
            than `max_elf_calories`. If it is then set `max_elf_calories`
            to `current_calories_total`. Then set `current_calories_total` to 0.
        */
        if (element == '') {
            max_elf_calories.push(current_calories_total);
            current_calories_total = 0;
        } else {
            current_calories_total += Number(element);
        }
    }

    // sort the `max_elf_calories` in reverse order (highest -> lowest)
    max_elf_calories.sort((a, b) => b - a);

    console.log(max_elf_calories.slice(0, 3).reduce((partialSum, a) => partialSum + a, 0));
}

part_1();
part_2();
